compiler_opt/rl/log_reader.py

```diff
@@ -0,0 +1,163 @@
+  the tensor.
+
+  Endianness is assumed to be the same as the log producer's.
+  """
+
+  def __init__(self, spec: TensorSpec, buffer: bytes):
+    self._spec = spec
+    self._buffer = buffer
+    # c_char_p is a nul-terminated string, so the more appropriate cast here
+    # would be POINTER(c_char), but unfortunately, c_char_p is the only
+    # type that can be constructed from a `bytes`. To capture our intent,
+    # we cast the c_char_p to
+    buffer_as_nul_ending_ptr = ctypes.c_char_p(self._buffer)
+    buffer_as_naked_ptr = ctypes.cast(buffer_as_nul_ending_ptr,
+                                      ctypes.POINTER(ctypes.c_char))
+    self._view = ctypes.cast(buffer_as_naked_ptr,
+                             ctypes.POINTER(self._spec.element_type))
+    self._len = math.prod(self._spec.shape)
+
+  def spec(self) -> TensorSpec:
+    return self._spec
+
+  def __len__(self) -> int:
+    return self._len
+
+  def __getitem__(self, index: int):
+    if index < 0 or index >= self._len:
+      raise IndexError(f'Index {index} out of range [0..{self._len})')
+    # pytype believes `index` is an object, despite all the annotations to the
+    # contrary.
+    return self._view[index]  # pytype:disable=wrong-arg-types
+
+
+@dataclasses.dataclass(frozen=True)
+class Header:
+  features: List[TensorSpec]
+  score: Optional[TensorSpec]
+
+
+def read_tensor(fs: BinaryIO, ts: TensorSpec) -> TensorValue:
+  size = math.prod(ts.shape) * ctypes.sizeof(ts.element_type)
+  data = fs.read(size)
+  return TensorValue(ts, data)
+
+
+def read_header(f: BinaryIO) -> Header:
+  header = json.loads(f.readline())
+  tensor_specs = [create_tensorspec(ts) for ts in header['features']]
+  score_spec = create_tensorspec(header['score']) if 'score' in header else None
+  return Header(features=tensor_specs, score=score_spec)
+
+
+def pretty_print_tensor_value(tv: TensorValue):
+  print(f'{tv.spec().name}: {",".join([str(v) for v in tv])}')
+
+
+@dataclasses.dataclass(frozen=True)
+class Record:
+  context: str
+  observation_id: int
+  feature_values: List[TensorValue]
+  score: Optional[TensorValue]
+
+
+def enumerate_log_from_stream(f: BinaryIO,
+                              header: Header) -> Generator[Record, None, None]:
+  """A generator that returns Record objects from a log file.
+
+  It is assumed the log file's header was read separately.
+  """
+  tensor_specs = header.features
+  score_spec = header.score
+  context = None
+  while event_str := f.readline():
+    event = json.loads(event_str)
+    if 'context' in event:
+      context = event['context']
+      continue
+    observation_id = int(event['observation'])
+    features = []
+    for ts in tensor_specs:
+      features.append(read_tensor(f, ts))
+    f.readline()
+    score = None
+    if score_spec is not None:
+      score_header = json.loads(f.readline())
+      assert int(score_header['outcome']) == observation_id
```

Comment:
what this assert mean?

Reply:
This should be more clear with the doc at the header of the file - the "outcome" and "observation" should have matching IDs, like so:
{"observation": 1}
<bunch of binary  data>
{"outcome": 1}
<the score>
...
{"observation": 2}
<bunch of binary  data>
{"outcome": 2}
<the score>

Just extra debugging/checking

Reply:
a broader question about the format design: why do we need the observation/outcome field? The sequence will naturally be the right observation sequence, right?

Reply:
Do you mean "why not just place the observation right after the tensor data"?

Two reasons:
- I want to use the same structure for the interactive mechanism, so this makes the logger output equivalent to the concatenated data the interactive solution would have. There, we need to clarify if some data coming is the outcome from the previous observation, or a new context
- We can now explore not having unnecessary observation records, for example for regalloc, just the last one matters. So a next step can have the observation record be optional (and the header have an indication that would be the case)

Reply:
can the context field tell whether it comes from previous observation or a new context?

and i think the current format needs to have score for all steps, right?

Reply:
it's not a context field, rather it's a json object that has a key labeled "context", and this will only appear when the context changes. Suppose we don't have an explicit json object saying "observation", and just dump the observation bytes. a reader would have to try and figure out if the new data is a new context by trying to parse it as json. That's not ideal.

The current format needs to have score for all steps, but this is something we can opt out of now - not in this patch, but the format easily allows it.

Reply:
then a more natural logging way may be:

// {"context": "aFunction"}
// {"observation": <bytes>}
// [optional] {"score": 0}
// {"observation": <bytes>}
// [optional] {"score": 0}
// {"observation": <bytes>}
// [optional] {"score": 0}
...

in this way you also don't need to have the "score" field in the header as parsing the key would tell that.

Reply:
The information in the header is how to parse the data: i.e. the tensor type and shape.

Reply:
It looks like the remaining discussion is on the format itself and in particular how the score signal is passed back. We're going to be iterating over this shortly - e.g. to add support for skipping non-existing partial rewards.

I plan to merge this later today so I can then tie training to it, and then dump protobuf support upstream. Let's discuss what we want to do to scoring in parallel, on this thread.

Reply:
sg, let's discuss the logging format offline.